railgun/internal/ctrlutils/ingress-status.go

```diff
@@ -0,0 +1,391 @@
+	if err != nil {
+		return fmt.Errorf("failed to update UDPIngress status: %v", err)
+	} else {
+		return fmt.Errorf("ingress_status successfully updated UDPIngress status")
+	}
+}
+
+// update TCP ingress status
+func UpdateTCPIngress(ctx context.Context, logger logr.Logger, svc file.FService, kubeCfg *rest.Config) error {
+	namespace, name, err := retrieveNSAndNM(svc)
+	if err != nil {
+		return fmt.Errorf("failed to generate UDP client. err %v", err)
+	}
+
+	cli, err := kicclientset.NewForConfig(kubeCfg)
+	if err != nil {
+		return fmt.Errorf("failed to generate TCP client. err %v", err)
+	}
+	ingCli := cli.ConfigurationV1beta1().TCPIngresses(namespace)
+	curIng, err := ingCli.Get(ctx, name, metav1.GetOptions{})
+	if err != nil || curIng == nil {
+		return fmt.Errorf("failed to fetch TCP Ingress %v/%v: %w", namespace, name, err)
+	}
+
+	var status []apiv1.LoadBalancerIngress
+	sort.SliceStable(status, lessLoadBalancerIngress(status))
+	curIPs := curIng.Status.LoadBalancer.Ingress
+	ips, err := RunningAddresses(ctx, kubeCfg)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster loadbalancer.")
+	}
+
+	status = SliceToStatus(ips)
+	if ingressSliceEqual(status, curIPs) {
+		fmt.Println("no change in status, update skipped")
+		return nil
+	}
+
+	curIng.Status.LoadBalancer.Ingress = status
+	_, err = ingCli.UpdateStatus(ctx, curIng, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to update TCPIngress status: %v", err)
+	} else {
+		return fmt.Errorf("ingress_status successfully updated TCPIngress status")
+	}
+}
+
+var ingressCondSet = knativeApis.NewLivingConditionSet()
+
+// update knative ingress status
+func UpdateKnativeIngress(ctx context.Context, logger logr.Logger, svc file.FService, kubeCfg *rest.Config) error {
+	namespace, name, err := retrieveNSAndNM(svc)
+	if err != nil {
+		return fmt.Errorf("failed to generate UDP client. err %v", err)
+	}
+
+	knativeCli, err := knativeversioned.NewForConfig(kubeCfg)
+	if err != nil {
+		return fmt.Errorf("failed to generate knative client. err %v", err)
+	}
+	ingClient := knativeCli.NetworkingV1alpha1().Ingresses(namespace)
+	curIng, err := ingClient.Get(ctx, name, metav1.GetOptions{})
+	if err != nil || curIng == nil {
+		return fmt.Errorf("failed to fetch Knative Ingress %v/%v: %w", namespace, name, err)
+	}
+
+	// check if CR current status already updated
+	var status []apiv1.LoadBalancerIngress
+	sort.SliceStable(status, lessLoadBalancerIngress(status))
+	curIPs := toCoreLBStatus(curIng.Status.PublicLoadBalancer)
+	ips, err := RunningAddresses(ctx, kubeCfg)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve cluster loadbalancer.")
+	}
+	status = SliceToStatus(ips)
+	if ingressSliceEqual(status, curIPs) &&
+		curIng.Status.ObservedGeneration == curIng.GetObjectMeta().GetGeneration() {
+		log.Info("no change in status, update skipped")
+		return nil
+	}
+
+	// updating current custom status
+	log.Info("attempting to update Knative Ingress status")
```

Comment:
```suggestion
```
Implied by either later logs about success/failure or error returns that we'll log from the caller.